crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+use std::io::{Cursor, Read};
+use std::sync::Arc;
+
+use itertools::Itertools;
+use zip::{read::ZipFile, result::ZipResult, ZipArchive};
+
+use crate::file_system::{FileRevision, FileSystem, FileSystemPath, FileType, Metadata, Result};
+
+#[derive(Debug, Clone)]
+pub struct VendoredFileSystem {
+    inner: VendoredZipArchive,
+}
+
+impl VendoredFileSystem {
+    const READ_ONLY_PERMISSIONS: u32 = 0o444;
+
+    pub fn new(raw_bytes: Arc<[u8]>) -> Result<Self> {
+        Ok(Self {
+            inner: VendoredZipArchive::new(raw_bytes)?,
+        })
+    }
+
+    /// Returns an owned copy of the underlying zip archive,
+    /// that can be queried for information regarding contained files and directories
+    fn vendored_archive(&self) -> VendoredZipArchive {
+        // Zip lookups require a mutable reference to the archive,
+        // which creates a nightmare with lifetimes unless you own the archive, hence the clone.
+        //
+        // The zip documentation states that:
+        //
+        //     At the moment, this type is cheap to clone if this is the case for the reader it uses.
+        //     However, this is not guaranteed by this crate and it may change in the future.
+        //
+        // So this should be fairly cheap *for now*.
+        self.inner.clone()
```

Comment:
I'm not sure if I want to rely on this behavior. It feels very easy to get wrong when updating the crate. 

I think what I would do is to wrap the `ZipArchive` in a mutex instead and require explicitly locking. This will reduce concurrency a bit but I don't expect that any of the zip operations take long.

Reply:
This is the problem I described as point (1) in my PR description. Because `ZipArchive::by_name` takes `&mut self`, it's very hard to implement `VendoredFileSystem::read()` (which must take `&self`). I think our options are as follows:
1. Creating a new `ZipArchive` instance every time `read()` is called
2. Using the existing `ZipArchive` instance every time `read()` is called, but cloning it to create an owned copy
3. Accept that `VendoredFileSystem::read()` will need to take `&mut self` rather than `&self` (I don't like this at all).
4. Unzip typeshed eagerly at runtime so that we don't need to use the `zip` crate at all to read vendored files.

Currently I'm doing Option (2) in this PR, but maybe I'm missing something.

Reply:
I switched the PR to use Option (6): "Something something interior mutability"